### ISSUE TITLE
Slow analysis time

### ISSUE THREAD

**Issue by erictraut** [microsoft/pyright#4950]
This is a tracking issue for a report in the pylance-release project.

For the original, refer to: https://github.com/microsoft/pylance-release/issues/4233

**Comment by erictraut** [comment]
With "python.analysis.useLibraryCodeForTypes" disabled, it takes 433ms to analyze the `brokerflow.py` file on my machine.

With "python.analysis.useLibraryCodeForTypes" enabled and with the latest version of pandas installed, it takes 10,038ms to analyze. Most of this time is in inferring types from the `sklearn` library. The most expensive are in the `cluster` and `decomposition` submodules within `sklearn`. With the pylance-bundled stubs for `sklearn`, the time drops to 2,409ms.

If I then install the `pandas-stubs` stub library, the time increases to 18,381ms! This is unexpected because stubs typical decrease analysis time. Disabling `useLibraryCodeForTypes` does not improve performance in this case. 

More investigation is needed here.

**Comment by erictraut** [comment]
I found the root cause of the performance issue. It was due to a method in the `WhaleRadar.fit` class. This method is atypical in that it does all of the following:
1. It uses instance variables that do not have annotated types. And in particular, it reads from these variables early in the function and assigns new values to these variables later in the method. That means the function effectively needs to be analyzed as a loop because the code at the end of the method affects the types earlier in the method. This requires evaluating many of the expressions multiple times.
2. It calls complex functions that return tuples with many entries, and it immediately unpacks these tuple values into many local variables and instance variables. This creates a bunch of dependencies between these various expressions that require significantly more work for the code flow engine to resolve. It results in a combinatoric increase in the number of times the code flow graph needs to be traversed to resolve all of these types.
3. In the middle of the method, several lambdas are used in a deeply-nested expression, and these lambdas reference instance variables and locals that are affected by 1 and 2.

Pyright needs to evaluate some expressions "speculatively" in cases where information that could affect the type of the expression is not yet known or may change. For example, if a lambda is passed as an argument to a function, the corresponding parameter type informs the "expected type" of the lambda, but the type of the parameter may be informed by later arguments (in cases where the function being called includes type variables in its parameter types). This means lambdas often need to be evaluated speculatively. Pyright retains a special "speculative type cache" in an attempt to avoid redundantly evaluating types in cases where their speculative dependencies do not change, but doing this for lambdas is difficult. Previously, it didn't retain cached values for any expressions evaluated speculatively for a lambda (or any of the expressions that the lambda referenced). This is what led to the poor performance in this particular case.

I figured out a way to safely cache the speculative types evaluated in the context of a lambda. This adds a bit of complexity to the caching mechanism, but it reduces the evaluation time in this particular file from about 18,400ms down to about 75ms on my machine (about a 250x speedup). This change probably won't have a broad impact on performance, but it could improve analysis times for other files that use lambdas in deeply nested expressions within complex functions.
